src/core/compiler/CircomCompilerFactory.ts

```diff
@@ -1,44 +1,138 @@
+import { CircomCompilerDownloader } from "./CircomCompilerDownloader";
+
+import { isVersionHigherOrEqual } from "./versioning";
+
+import { CompilerInfo, CompilerPlatformBinary, ICircomCompiler, NativeCompiler } from "../../types/core";
 
 // eslint-disable-next-line
 const { Context } = require("@distributedlab/circom2");
 
 export class CircomCompilerFactory {
-  public static createWASMCircomCompiler(version: CompilerVersion): IWASMCircomCompiler {
-    switch (version) {
-      case "0.2.18":
-        return new WASMCircomCompiler(this._getCircomCompiler("@distributedlab/circom2/circom.wasm"));
-      default:
-        throw new HardhatZKitError(`Unsupported Circom compiler version - ${version}. Please provide another version.`);
+  private static instance: CircomCompilerFactory;
+
+  private constructor() {}
+
+  public static getInstance(): CircomCompilerFactory {
+    if (!CircomCompilerFactory.instance) {
+      CircomCompilerFactory.instance = new CircomCompilerFactory();
     }
+
+    return CircomCompilerFactory.instance;
   }
 
-  public static async createNativeCircomCompiler(version: CompilerVersion): Promise<ICircomCompiler> {
-    switch (version) {
-      case "0.2.18":
-        return new NativeCircomCompiler();
-      default:
-        throw new HardhatZKitError(`Unsupported Circom compiler version - ${version}. Please provide another version.`);
+  public async createBinaryCircomCompiler(version: string, isVersionStrict: boolean): Promise<ICircomCompiler> {
+    if (!isVersionHigherOrEqual(LATEST_SUPPORTED_CIRCOM_VERSION, version)) {
+      throw new HardhatZKitError(`Unsupported Circom compiler version - ${version}. Please provide another version.`);
     }
-  }
 
-  public static async checkNativeCompilerExistence() {
     try {
-      await execCall("circom", ["--version"]);
-    } catch (e) {
-      throw new HardhatZKitError(
-        "Native Circom compiler was not found. Set the compiler globally or change the 'nativeCompiler' flag to false",
-      );
+      const nativeCompiler = await this._getNativeCompiler();
+
+      const isValidVersion = isVersionStrict
+        ? nativeCompiler.version === version
+        : isVersionHigherOrEqual(nativeCompiler.version, version);
+
+      if (isValidVersion) {
+        Reporter!.reportCompilerVersion(nativeCompiler.version);
+
+        return new BinaryCircomCompiler(nativeCompiler.binaryPath);
+      }
+    } catch (error: any) {
+      Reporter!.verboseLog("compiler-creation", `No native compiler compatible with version ${version} found`);
+    }
+
+    let compilerPlatformBinary: CompilerPlatformBinary;
+    if (
+      isVersionStrict &&
+      os.arch() === "arm64" &&
+      !isVersionHigherOrEqual(version, OLDEST_SUPPORTED_ARM_CIRCOM_VERSION)
+    ) {
+      compilerPlatformBinary = CircomCompilerDownloader.getCompilerPlatformBinary("x64");
+    } else {
+      compilerPlatformBinary = CircomCompilerDownloader.getCompilerPlatformBinary();
+    }
```

Comment:
Let's create a separate private function for this. Maybe `resolveCompilerPlatformForARM()`?